enhancements/ingress/gateway-api-with-cluster-ingress-operator.md

```diff
@@ -0,0 +1,434 @@
+"istio-system" namespace.  To ensure backwards compatibility, the Ingress
+Operator must do one of the following:
+
+* Use the same name and namespace: "istio-system/basic".
+* Migrate an existing SMCP from "istio-system" to "openshift-ingress".
+* Provide an API to tell the Ingress Operator what SMCP to use.
+
+Placement, possible migration, and ownership of the ServiceMeshControlPlane need
+to be clarified in this enhancement.  This is an open design question.
+
+#### Managing the ServiceMeshMemberRoll
+
+OSSM 2.3 only watches resources in namespaces specified in an
+ServiceMeshMemberRoll CR associated with the ServiceMeshControlPlane.  We hope
+that this limitation will be removed from OSSM in time for this enhancement;
+this is an open release-planning question.  Otherwise, the Ingress Operator or
+cluster admin must add each namespace with HTTPRoute objects that Istiod should
+reconcile to the ServiceMeshMemberRoll, and the approach we take and ownership
+of ServiceMeshMemberRoll need to be clarified in this enhancement; this is an
+open design question.
+
+#### Gateway API v1alpha1 CRDs
+
+Istiod fails to become ready if the expected CRDs do not all exist.  OSSM 2.3 is
+based on a version of Istio that expects the Gateway API v1alpha1 CRDs to exist.
+In particular, Gateway API v1alpha1 includes the following CRDs that are not
+defined in Gateway API v1beta1: tcproutes, tlsroutes, and udproutes.  For the
+time being, this means that it is necessary to install the v1alpha1 CRDs in
+order for OSSM 2.3 to function. We hope that OSSM will have a new release based
+on a newer version of Istio in time for this enhancement. This is an open
+release-planning question.
+
+#### Security Policy
+
+TBD.
+
+### Risks and Mitigations
+
+#### Automatic deployments
+
+OpenShift Service Mesh and Istio have a feature called [automated
+deployment](https://istio.io/latest/docs/tasks/traffic-management/ingress/gateway-api/#automated-deployment)
+that creates an Envoy deployment and service for each Gateway.  Enabling
+automated deployments and also allowing arbitrary users to create Gateways would
+create a new attack surface for untrusted cluster users.  For this reason, the
+default policy allows only cluster admins to create Gateways.
+
+### Drawbacks
+
+OpenShift router is a robust, mature, feature-rich, well understood ingress
+solution based on an extremely performant proxy (HAProxy).  In contrast,
+OpenShift Service Mesh, Istio, and Envoy are disadvantaged in all these
+respects.  However, OpenShift router and OpenShift Service Mesh can co-exist on
+the same cluster, and OpenShift router remains the default ingress solution for
+the time being.  Moreover, Istio and Envoy enjoy rapid community adoption, as
+does Gateway API.  Using Istio's Gateway API implementation aligns OpenShift's
+ingress solution with its mesh solution with both using Istio/Envoy and also
+aligns with the community uptake of Gateway API as the standard lowercase-"i"
+ingress API.
```

Comment:
I'm sure there will be drawbacks to talk about regarding unified control plane once we get a more finalized design (i.e. I'm willing to bet C-I-O has to make assumptions when installing ServiceMeshControlPlane or how to use an existing ServiceMeshControlPlane), but I'll just leave this here as a reminder.

Also there might be drawbacks of using the same operator to accomplish both Route and Gateway API.

Reply:
Yeah, we'll probably find more challenges with having a unified control-plane as we go.  

I don't think using the same operator is a drawback.  At a conceptual level, cluster-ingress-operator is what the name says: it's the operator that manages ingress for the cluster.  There's no reason it cannot have one controller (or set of controllers) for IngressController and another controller (or set of controllers) for Gateway API, and there is enough commonality (e.g. DNS management) that I believe using the same operator is a significant advantage.